### PR TITLE
new approach to featuregates for coordination in the cluster

### PR DESCRIPTION
**Pull request by deads2k**
This enhancement aims to reduce the effort required to add a feature gate to TechPreviewNoUpgrade and to promote
that feature gate to Default.
Feature gates in OpenShift are enabled and disabled in a particular FeatureSet, mixing and matching is not allowed.
Prior to this enhancement, it is necessary to vendor openshift/api into every impacted repository.
After this enhancement, it will only be necessary to vendor into cluster-config-operator.

cc @openshift/openshift-staff-engineers @JoelSpeed

### CODE REVIEW

**Review by JoelSpeed — COMMENTED** [review]
Looks great, just got one question about how we are going to test feature promotion

File: dev-guide/feature-gate.md [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+
+### User Stories
+
+As a staff engineer or release manager, I want to have no-developer-action result in features of unproven reliability
+inaccessible-by-default in long-lived clusters.
+
+As a developer, I want to have a low friction way to add a feature gate to TechPreviewNoUpgrade and to promote a feature
+gate to on-by-default.
+
+As an API approver, I want to have evidence of feature reliability before promoting feature gates to on-by-default.
```
Comment (bparees) [inline dev-guide/feature-gate.md:39]: it's a little weird to me that this implies it's an api approver decision.  I see that as a happenstance of where featuregates live, but i wouldn't expect api approvers to make the decision and i'd expect api approvers to respect it when it is made, not go examine the evidence and second guess it.  (that's as much about "they aren't the experts here" as it is "they don't have time to be doing that")

which begs an important question:  who *should* be making the decision?  
TRT?
a staff engineer?
the dev team?
QE?
some combination thereof?

File: dev-guide/feature-gate.md [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+
+1. have a low friction way to add a feature gate to TechPreviewNoUpgrade
+2. have to promote a feature gate to on-by-default
+3. allow feature gates to move from TechPreviewNoUpgrade to Default without releases as TechPreviewNoUpgrade first.
+
+### Non-Goals
+
+1. Change thresholds for requiring TechPreviewNoUpgrade.
+   This is a possible future goal, but an update here would be required.
+2. Change thresholds for promoting from TechPreviewNoUpgrade to Default.
```
Comment (bparees) [inline dev-guide/feature-gate.md:51]: do we feel we have a well defined existing threshold?  if not, this non-goal seems to conflict w/ the user story above of "i have sufficient evidence to promote this to on by default"  (I don't know how we meet that user story w/o defining those thresholds, so if they aren't well defined then by implication they need to be changed)

File: dev-guide/feature-gate.md [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+is now a benefit of using these capabilities to test our code before impacting real clusters.
+
+
+To overcome the original limitation, we will need to key the enabled/disabled feature gates with the payload version
+that they are for.
+This will ensure that older versions do not enable feature gates that were not GA until the later version.
+To do this we will
+1. Update FeatureGateStatus to contain a list of enabled and disabled feature gates for up to every version listed
+   in the CVO history.
+2. Update cluster-config-operator to have a control loop owned by api-approvers to set the FeatureGateStatus
```
Comment (bparees) [inline dev-guide/feature-gate.md:90]: again not really sure why would be owned by api-approvers.

I see api review/approval as a distinct need/skillset/requirement that is unrelated to "definition of the maturity of a feature"

is it just the lack of another obvious team/set of people to own this code?

File: dev-guide/feature-gate.md [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+3. Update library-go to make it easy to
+   1. wait for FeatureGates to be read from the cluster for this version
+   2. make it easy to function with a development build on a patched cluster
+   3. make it easy to read the current state of FeatureGates
+   4. by default, exit when feature gates change (most processes don't react cleanly to changes at runtime)
```
Comment (bparees) [inline dev-guide/feature-gate.md:95]: this is going to result in extra+stampeding restarts (of processes, not nodes and not pods, i realize) on upgrades, right?

i know we have a semi-common pattern of self-terminating processes to pick up changes, but this seems particularly widespread and likely to happen in close temporal proximity.  Maybe not a concern, but felt worthy of note/discussion.

wondering if there are optimizations to avoid this.

e.g. there's no reason for the process to restart if the only hange in the featuregates is to versions they are ignoring anyway

so if we can update the featuregates (add the new version entry) before we upgrade the various process images, then when they restart due to the image change, they'll come up, see the latest feature gate values, and we can avoid any extra restarts.

File: dev-guide/feature-gate.md [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+   5. make it easy to wire a hook to handle FeatureGate updates if exit isn't desired
+4. Update existing operators to use the library-go implementation to consume feature gates in this way
+5. Update the installer to pass full FeatureGate manifests to the five-ish operators that participate in rendering
+   This is needed to ensure that cluster-config-operator is updated, all the bootstrap processes will have the same
```
Comment (bparees) [inline dev-guide/feature-gate.md:99]:
```suggestion
   This is needed to ensure that when cluster-config-operator is updated, all the bootstrap processes will have the same
```

?

File: dev-guide/feature-gate.md [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+Clusters will not install without the controller functional and we don't support changing feature gates back from
+TechPreviewNoUpgrade, so e2e tests are impractical.
+
+### Graduation Criteria
+
+#### Dev Preview -> Tech Preview
+
+#### Tech Preview -> GA
+
+Since this is actually controlling the gates, it is not practical to pre-test it.
```
Comment (bparees) [inline dev-guide/feature-gate.md:211]: fair enough, but fundamentally the GA criteria for this are "all our operators that read featuregates, are reading them using this new mechanism", right?

If any code is out there still vendoring openshift/api in order to understand which gates are associated with which featureset, then we're at risk of ending up out of sync somewhere.

I assume your list below more or less covers the places that need to be touched, but it seems like a useful "acceptance criteria" to note.

File: dev-guide/feature-gate.md (hunk 1 above)
Comment (deads2k) [reply dev-guide/feature-gate.md:39]: >  i wouldn't expect api approvers to make the decision and i'd expect api approvers to respect it when it is made

Similar to, "you need a merged enhancement" (api-approvers don't approve the enhancement), I would like to see, "you need reliability data" as a way to demonstrate, "yeah, basic threshold met so you can merge".  


> which begs an important question: who should be making the decision?

Do you see a reason to change the mechanics of how these are approved as part of this enhancement describing how they are consumed?  I'd rather debate mechanics of approval separately.

My final desire: bot does the merge by looking at the component reliability data.  My this year desire: staff eng decides by looking at the component reliability data.  My immediate goal: api-approvers aren't allowed to merge until some kind of feature reliability is presented.

File: dev-guide/feature-gate.md (hunk 2 above)
Comment (deads2k) [reply dev-guide/feature-gate.md:51]: It's been informal so far.  I'm willing to describe in more detail, but to my knowledge, the informal, existing mechanism has been working well and I wasn't inclined to change it.

File: dev-guide/feature-gate.md [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+This enhancement aims to reduce the effort required to add a feature gate to TechPreviewNoUpgrade and to promote
+that feature gate to Default.
+Feature gates in OpenShift are enabled and disabled in a particular FeatureSet, mixing and matching is not allowed.
+Prior to this enhancement, it is necessary to vendor openshift/api into every impacted repository.
+After this enhancement, it will only be necessary to vendor into cluster-config-operator.
+
+## Motivation
+
+Reduce the effort required to add and promote feature gates, to encourage and eventually require using
+feature gates to introduce new features into OCP.
```
Comment (dhellmann) [inline dev-guide/feature-gate.md:29]: I like the idea of making them easier to use, but I don't support the end goal of requiring feature gates. We should have the conversation about whether we will do that separately from this enhancement.

File: dev-guide/feature-gate.md [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+
+### User Stories
+
+As a staff engineer or release manager, I want to have no-developer-action result in features of unproven reliability
+inaccessible-by-default in long-lived clusters.
```
[the remark this answers is not in the document]
Comment (dhellmann) [reply dev-guide/feature-gate.md:34]: I don't understand this sentence. What is being prevented or allowed and why does a staff engineer or release manager care more about that than any other contributor?

File: dev-guide/feature-gate.md (hunk 3 above)
Comment (deads2k) [reply dev-guide/feature-gate.md:90]: > is it just the lack of another obvious team/set of people to own this code?

Yes.  And avoiding the dumping ground of unowned cruft that was developing six months back or so.  Small quantity of code to bootstrap a configuration API owned by the people approving those configuration APIs.

File: dev-guide/feature-gate.md [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+
+As a developer, I want to have a low friction way to add a feature gate to TechPreviewNoUpgrade and to promote a feature
+gate to on-by-default.
+
+As an API approver, I want to have evidence of feature reliability before promoting feature gates to on-by-default. 
+
+### Goals
+
+1. have a low friction way to add a feature gate to TechPreviewNoUpgrade
+2. have to promote a feature gate to on-by-default
```
Comment (dhellmann) [inline dev-guide/feature-gate.md:44]: Should this read "have a low friction way to..."?

File: dev-guide/feature-gate.md [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+4. Nag api-approvers with link to your PR right away and then every 24h or so until they merge it.
+
+Notice that this flow eliminates vendoring into multiple repositories and makes it possible to test a feature gate change
+with a single PR, so /payload testing functions properly.
+
+### Phase 2
+1. Open PR to openshift/api to add your feature gate to [TechPreviewNoUpgrade](https://github.com/openshift/api/blob/master/config/v1/types_feature.go#L117).
+   The PR should be confined to just the feature gate change and should include a link to a merged enhancement.
+2. Nag api-approvers with link to your PR right away and then every 24h or so until they merge it.
+3. Automation vendors openshift/api into cluster-config-operator in a few hours.
```
[the remark this answers is not in the document]
Comment (dhellmann) [reply dev-guide/feature-gate.md:70]: If the bot is only opening a PR, then step 4 from above still seems necessary to get it merged.

File: dev-guide/feature-gate.md (hunk 1 above)
Comment (dhellmann) [reply dev-guide/feature-gate.md:39]: The mechanics of approval are my primary concern with this, so either we need to address that first or as part of this. We cannot create a system in which the forward progress of more than 370 people spread across all of the teams impacted by this change depends on the review capacity of 2 API reviewers or even 13 staff engineers.

File: dev-guide/feature-gate.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (bparees) [reply dev-guide/feature-gate.md:34]: I struggled to parse it also, but i believe the intended meaning is:

i want unproven/verified features(code) to be inert by default.  they can be non-inert in techpreview/non-upgradeable clusters, i.e. clusters w/ techpreviewnoupgrade enabled, and they can be non-inert once reliability of the feature has been proven.

but yes, as for who the user that cares is in this story, it really should be all  of us (including customers), but release managers in particular benefit in the sense that they are less likely to hit a situation where we need an urgent bug fix or reversion in order to be able to ship on time.  (but again, really everyone benefits from that lack of firedrills)

File: dev-guide/feature-gate.md (hunk 6 above)
Comment (deads2k) [reply dev-guide/feature-gate.md:210]: clarified in doc.

File: dev-guide/feature-gate.md (hunk 1 above)
Comment (deads2k) [reply dev-guide/feature-gate.md]: struck sentence. This is informal today and will not go away.  But changing it isn't a goal.  See pre-existing non-goal 2 on this list.

File: dev-guide/feature-gate.md [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+last-updated: 2023-03-29
+---
+
+# FeatureGates, round 2
+
+## Summary
+
+This enhancement aims to reduce the effort required to add a feature gate to TechPreviewNoUpgrade and to promote
+that feature gate to Default.
+Feature gates in OpenShift are enabled and disabled in a particular FeatureSet, mixing and matching is not allowed.
```
[the remark this answers is not in the document]
Comment (deads2k) [reply dev-guide/feature-gate.md:22]: added a section at the bottom for random questions like this.

File: dev-guide/feature-gate.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (deads2k) [reply dev-guide/feature-gate.md]: clarified in doc

File: dev-guide/feature-gate.md (hunk 7 above)
Comment (deads2k) [reply dev-guide/feature-gate.md:29]: I can strike the sentence (this enhancement is about making it easier to use them), but I included it for transparency about my objective.  If it prevents you from agreeing to improve the current state of affairs I can remove it.

**Comment by deads2k** [comment]
Removed contentious bits and repushed.

File: dev-guide/feature-gate.md (hunk 7 above)
Comment (deads2k) [reply dev-guide/feature-gate.md]: Struck. It's distracting.  Let's talk in a year.

File: dev-guide/feature-gate.md (hunk 9 above)
Comment (deads2k) [reply dev-guide/feature-gate.md]: fixed.

File: dev-guide/feature-gate.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (bparees) [reply dev-guide/feature-gate.md:22]: thanks

File: dev-guide/feature-gate.md [hunk 12]
```diff
@@ -0,0 +1,293 @@
...
+TechPreviewNoUpgrade, so e2e tests are impractical.
+
+### Graduation Criteria
+
+#### Dev Preview -> Tech Preview
+
+#### Tech Preview -> GA
+
+Since this is actually controlling the gates, it is not practical to pre-test it.
+Once unit tests pass and clusters successfully install equivalently using this mechanism, PRs will be opened against
```
Comment (rphillips) [inline dev-guide/feature-gate.md:210]: I am interested how the MCO and conversely the Kubelet would digest the configured FeatureGates at bootstrap.

MCO Reference for Features Gates: https://github.com/openshift/machine-config-operator/pull/2668

File: dev-guide/feature-gate.md (hunk 4 above)
Comment (JoelSpeed) [reply dev-guide/feature-gate.md:93]: > so if we can update the featuregates (add the new version entry) before we upgrade the various process images, then when they restart due to the image change, they'll come up, see the latest feature gate values, and we can avoid any extra restarts.

The config operator today is run level 30, which for context is the same as MAPI and comes after most of the control plane components. If we were to promote the run level, then the new version feature gate information should be available early enough in the upgrade to avoid the duplicate restarts of all the operators.

I agree the default works so maybe this is something we want to trial and then optimise later?

> Remember, this is the operator, not the operand.

What are the odds that various components bake this directly into the operand? I could see it in machine API for example, if there's an AWS specific feature, it may be more obvious to put the feature gate processing into the AWS implementation rather than using a flag controlled by the operator.

Maybe we want to clarify in the documentation for this process that operands should use feature flags controlled by operators to avoid the exits/need to gracefully handle changes

File: dev-guide/feature-gate.md [hunk 13]
```diff
@@ -0,0 +1,293 @@
...
+We're done when all operators using feature gates are reading from FeatureGate.Status instead of vendoring the list.
+
+#### Removing a deprecated feature
+
+- Announce deprecation and support policy of the existing feature
+- Deprecate the feature
+
+### Upgrade / Downgrade Strategy
+
+On upgrade to the first level with this change, the cluster-config-operator goes first, so the FeatureGateStatus will
```
Comment (JoelSpeed) [inline dev-guide/feature-gate.md:236]: Is this true? I just double checked and the CCO deployment manifest is at run level 30, so I wouldn't expect it to be upgraded until after all of the core control plane components

File: dev-guide/feature-gate.md [hunk 14]
```diff
@@ -0,0 +1,293 @@
...
+   This is a possible future goal, but an update here would be required.
+2. Change thresholds for promoting from TechPreviewNoUpgrade to Default.
+   This is a possible future goal, but an update here would be required.
+
+## Proposal
+### Phase 1
+For developers using this, the first phase will look like
+1. Open PR to openshift/api to add your feature gate to [TechPreviewNoUpgrade](https://github.com/openshift/api/blob/master/config/v1/types_feature.go#L117).
+   The PR should be confined to just the feature gate change and should include a link to a merged enhancement.
+2. Nag api-approvers with link to your PR right away and then every 24h or so until they merge it.
```
Comment (JoelSpeed) [inline dev-guide/feature-gate.md:56]: Is it worth noting that nagging API approvers should be in `#forum-api-review` to balance the load rather than going directly to `<insert your favourite API approver here>` in a DM?

Should we have a slack alias to use to allow the group to be summoned to a particular thread?

File: dev-guide/feature-gate.md [hunk 15]
```diff
@@ -0,0 +1,293 @@
...
+	// Only featureGates with .version in the ClusterVersion.status will be present in this list.
+	// +listType=map
+	// +listMapKey=version
+	FeatureGates []FeatureGateDetails `json:"featureGates"`
+}
+
+type FeatureGateDetails struct {
+	// version matches the version provided by the ClusterVersion and in the ClusterOperator.Status.Versions field.
+	// +kubebuilder:validation:Required
+	// +required
```
Comment (JoelSpeed) [inline dev-guide/feature-gate.md:128]: Which tools actually still use this? I thought we had a preference to just use `+kubebuilder:validation:Required`?

File: dev-guide/feature-gate.md [hunk 16]
```diff
@@ -0,0 +1,293 @@
...
+	// +optional
+	Enabled []FeatureGateAttributes `json:"enabled"`
+	// disabled is a list of all feature gates that are disabled in the cluster for the named version
+	// +optional
+	Disabled []FeatureGateAttributes `json:"disabled"`
+}
+
+type FeatureGateAttributes struct {
+	// name is the name of the FeatureGate
+	// +kubebuilder:validation:Pattern=`^([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.?$`
```
Comment (JoelSpeed) [inline dev-guide/feature-gate.md:140]: Do we have any validation at the point that feature gates are added to the API that matches this? I don't see it in the featuregates API but I know there's an admission plugin for that which may be doing the validation

File: dev-guide/feature-gate.md [hunk 17]
```diff
@@ -0,0 +1,293 @@
...
+After this enhancement, it will only be necessary to vendor into cluster-config-operator.
+
+## Motivation
+
+Reduce the effort required to add and promote feature gates.
+
+### User Stories
+
+As a staff engineer or release manager, I want to have no-developer-action result in features of unproven reliability
+inaccessible-by-default in upgradeable clusters.
```
[the remark this answers is not in the document]
Comment (sdodson) [reply dev-guide/feature-gate.md:33]: @deads2k Maybe this should've been "accessible-by-default" ? You want all features to go through a tech-preview phase where they're expected to show that their tests pass at historic product wide baseline before they're available without TechPreviewNoUpgrade featureset.

File: dev-guide/feature-gate.md (hunk 14 above)
Comment (dhellmann) [reply dev-guide/feature-gate.md:56]: We talked about this in the staff engineer call today and agreed to move ahead with the technical aspects of this to unblock some other work, and have a deeper conversation about delegation of the process portion in the future.

File: dev-guide/feature-gate.md [hunk 18]
```diff
@@ -0,0 +1,293 @@
...
+TechPreviewNoUpgrade CI jobs, stabilized them, and developed means of inspecting reliability of tests and jobs, there
+is now a benefit of using these capabilities to test our code before impacting real clusters.
+
+
+To overcome the original limitation, we will need to key the enabled/disabled feature gates with the payload version
+that they are for.
+This will ensure that older versions do not enable feature gates that were not GA until the later version.
+To do this we will
+1. Update FeatureGateStatus to contain a list of enabled and disabled feature gates for up to every version listed
+   in the CVO history.
```
Comment (wking) [inline dev-guide/feature-gate.md:87]: ClusterVersion `status.history` is not exhaustive: openshift/cluster-version-operator#791, openshift/cluster-version-operator#805, #1153.  It's possible that the fraction of history maintained by #1153 is sufficient for this enhancement.  And it's possible that implicit in this enhancement is that `status.featureGates` will be pruned to keep up with ClusterVersion `status.history` pruning.

I'm not all that clear on why you'd want such deep history in `featureGates` though.  Wouldn't the most recent `Completed` version, and any subsequent `Partial`  versions be sufficient?  Or possibly the useful version horizon extends back to the previous minor version, to accommodate possible eventual support for [rollbacks within a z stream](https://issues.redhat.com/browse/OTA-455)?  But once you've `Completed`  a 4.y.z, I don't understand why any 4.(y-1) or older feature gate information would matter.

**Comment by sdodson** [comment]
/remove-lifecycle stale
My understanding is that this is actually essentially implemented at this point. Perhaps a quick audit for accuracy with current implementation and then merge?

**Comment by JoelSpeed** [comment]
/reopen
/remove-lifecycle rotten

Yeah this is basically complete, we should review and merge for future us to have a record of what we did

File: dev-guide/feature-gate.md [hunk 19]
```diff
@@ -0,0 +1,293 @@
...
+4. Nag api-approvers with link to your PR right away and then every 24h or so until they merge it.
+
+Notice that this flow eliminates vendoring into multiple repositories and makes it possible to test a feature gate change
+with a single PR, so /payload testing functions properly.
+
+### Phase 2
+1. Open PR to openshift/api to add your feature gate to [TechPreviewNoUpgrade](https://github.com/openshift/api/blob/master/config/v1/types_feature.go#L117).
+   The PR should be confined to just the feature gate change and should include a link to a merged enhancement.
+2. Nag api-approvers with link to your PR right away and then every 24h or so until they merge it.
+3. Automation vendors openshift/api into cluster-config-operator and opens a PR in a few hours.
```
Comment (JoelSpeed) [inline dev-guide/feature-gate.md:67]: We haven't done this bit yet, is anyone prioritising it?

**Comment by dhellmann** [comment]
(automated message) This pull request is closed with lifecycle/rotten. It does not appear to be linked to a valid Jira ticket. Should the PR be reopened, updated, and merged? If not, removing the lifecycle/rotten label will tell this bot to ignore it in the future.

**Comment by dhellmann** [comment]
(automated message) This pull request is closed with lifecycle/rotten. It does not appear to be linked to a valid Jira ticket. Should the PR be reopened, updated, and merged? If not, removing the lifecycle/rotten label will tell this bot to ignore it in the future.

**Comment by dhellmann** [comment]
(automated message) This pull request is closed with lifecycle/rotten. It does not appear to be linked to a valid Jira ticket. Should the PR be reopened, updated, and merged? If not, removing the lifecycle/rotten label will tell this bot to ignore it in the future.
